File: test_mcp_servers.py
#!/usr/bin/env python3
"""
Test script for MCP Servers
"""
import asyncio
import json
import subprocess
import sys
from typing import Dict, Any

async def test_mcp_server(package_name: str, server_name: str) -> Dict[str, Any]:
    """Test if an MCP server package can be executed"""
    try:
        # Try to run the MCP server with help/version flag
        result = subprocess.run(
            ["npx", "-y", package_name, "--help"],
            capture_output=True,
            text=True,
            timeout=10
        )
        return {
            "server": server_name,
            "package": package_name,
            "status": "success" if result.returncode == 0 else "error",
            "returncode": result.returncode,
            "stdout": result.stdout[:500],  # Limit output
            "stderr": result.stderr[:500]
        }
    except subprocess.TimeoutExpired:
        return {
            "server": server_name,
            "package": package_name,
            "status": "timeout",
            "error": "Command timed out"
        }
    except Exception as e:
        return {
            "server": server_name,
            "package": package_name,
            "status": "error",
            "error": str(e)
        }

async def test_all_mcp_servers():
    """Test all configured MCP servers"""
    # Read MCP config
    try:
        with open('mcp_config.json', 'r') as f:
            config = json.load(f)
    except Exception as e:
        print(f"Error reading mcp_config.json: {e}")
        return

    servers = config.get('mcpServers', {})
    results = []

    print("Testing MCP Servers...")
    print("=" * 50)

    for server_name, server_config in servers.items():
        if not server_config.get('enabled', False):
            print(f"Skipping disabled server: {server_name}")
            continue

        package_name = server_config['args'][1]  # Extract package name from args
        print(f"\nTesting {server_name} ({package_name})...")

        result = await test_mcp_server(package_name, server_name)
        results.append(result)

        print(f"  Status: {result['status']}")
        if result['status'] == 'success':
            print("  ✓ Server responded successfully")
        elif result['status'] == 'timeout':
            print("  ! Server timed out")
        elif result['status'] == 'error':
            print(f"  ✗ Error: {result.get('error', 'Unknown error')}")

    # Summary
    print("\n" + "=" * 50)
    print("SUMMARY:")
    success_count = sum(1 for r in results if r['status'] == 'success')
    print(f"Total servers tested: {len(results)}")
    print(f"Successful: {success_count}")
    print(f"Failed: {len(results) - success_count}")

    if success_count > 0:
        print("\n✓ MCP servers are properly installed and accessible!")
        print("You can now use these tools in your AI assistant.")
    else:
        print("\n✗ No MCP servers working. Check installation and configuration.")

if __name__ == "__main__":
    asyncio.run(test_all_mcp_servers())

File: test_mcp_servers.js
#!/usr/bin/env node

/**
 * MCP Server Validation Script
 * Tests each enhanced MCP server to ensure they work correctly
 */

import { spawn } from 'child_process';
import path from 'path';
import fs from 'fs';

const servers = [
  {
    name: 'gpt-researcher-mcp',
    path: 'docker/mcp-servers/advanced/gptr-mcp/index.js',
    testInput: 'Analyze DFS player correlations',
  },
  {
    name: 'serena-code-analysis',
    path: 'docker/mcp-servers/advanced/serena/src/server.js',
    testInput: 'apps/api-python/optimization_engine.py',
  },
  {
    name: 'claude-flow',
    path: 'docker/mcp-servers/advanced/claude-flow-mcp/server.js',
    testInput: 'Optimize DFS lineup generation',
  },
  {
    name: 'google-genai-toolbox',
    path: 'docker/mcp-servers/advanced/genai-toolbox/server.js',
    testInput: 'Predict player performance trends',
  },
  {
    name: 'pipedream-chat',
    path: 'docker/mcp-servers/advanced/pipedream-mcp/server.js',
    testInput: 'Fetch latest NFL injury data',
  },
];

async function testServer(server) {
  console.log(`\n🧪 Testing ${server.name}...`);

  return new Promise((resolve, reject) => {
    const absolutePath = path.resolve(server.path);
    console.log(`   Path: ${absolutePath}`);

    // Check if file exists
    if (!fs.existsSync(absolutePath)) {
      console.log(`   ❌ FILE NOT FOUND: ${absolutePath}`);
      resolve({
        server: server.name,
        status: 'FAILED',
        reason: 'File not found',
        path: absolutePath,
      });
      return;
    }

    console.log(`   ✅ File exists`);

    // Try to start the server briefly to test syntax/loading
    const timeout = setTimeout(() => {
      if (child) {
        child.kill();
        console.log(`   ✅ Server started successfully (${server.name})`);
        resolve({
          server: server.name,
          status: 'SUCCESS',
          reason: 'Started without errors',
        });
      }
    }, 2000);

    let child;
    try {
      child = spawn('node', [absolutePath], {
        stdio: ['pipe', 'pipe', 'pipe'],
        cwd: process.cwd(),
        timeout: 3000,
      });

      let startedSuccessfully = false;
      let errorOutput = '';

      child.stderr.on('data', data => {
        errorOutput += data.toString();
        // Look for successful startup indicators
        if (
          errorOutput.includes('server running') ||
          errorOutput.includes('listening') ||
          !errorOutput.includes('Error')
        ) {
          startedSuccessfully = true;
        }
      });

      child.stdout.on('data', data => {
        const output = data.toString();
        if (output.includes('server running') || output.includes('listening')) {
          startedSuccessfully = true;
        }
      });

      child.on('error', error => {
        clearTimeout(timeout);
        console.log(`   ❌ FAILED: ${error.message}`);
        resolve({ server: server.name, status: 'FAILED', reason: error.message });
      });

      child.on('exit', (code, signal) => {
        clearTimeout(timeout);
        if (code === 0 || startedSuccessfully) {
          console.log(`   ✅ Server exited cleanly (${server.name})`);
          resolve({ server: server.name, status: 'SUCCESS', reason: 'Clean exit' });
        } else {
          console.log(`   ❌ Server failed with code ${code}`);
          resolve({
            server: server.name,
            status: 'FAILED',
            reason: `Exit code ${code}`,
            stderr: errorOutput,
          });
        }
      });
    } catch (error) {
      clearTimeout(timeout);
      console.log(`   ❌ SPAWN ERROR: ${error.message}`);
      resolve({ server: server.name, status: 'FAILED', reason: error.message });
    }
  });
}

async function runTests() {
  console.log('🚀 Starting MCP Server Validation...');
  console.log('=====================================');

  const results = [];

  for (const server of servers) {
    const result = await testServer(server);
    results.push(result);
  }

  console.log('\n📊 VALIDATION RESULTS');
  console.log('=====================');

  let passCount = 0;
  let failCount = 0;

  results.forEach(result => {
    if (result.status === 'SUCCESS') {
      console.log(`✅ ${result.server}: ${result.reason}`);
      passCount++;
    } else {
      console.log(`❌ ${result.server}: ${result.reason}`);
      failCount++;
    }
  });

  console.log('\n🎯 SUMMARY');
  console.log(`✅ Passing: ${passCount}`);
  console.log(`❌ Failing: ${failCount}`);
  console.log(
    `📈 Success Rate: ${((passCount / (passCount + failCount)) * 100).toFixed(1)}%`
  );

  if (failCount === 0) {
    console.log('\n🎉 ALL TESTS PASSED! All MCP servers are working correctly.');
  } else {
    console.log('\n⚠️  SOME TESTS FAILED. Review FAILED servers above.');
  }

  return results;
}

// Run tests automatically when executed
await runTests().catch(console.error);
